### PR TITLE
Kubelet: filter out terminated pods in SyncPods

### PR DESCRIPTION
**Pull request by yujuhong**
Once a pod reaches a terminated state (whether failed or succeeded), it should
not transit out ever again. Currently, kubelet relies on examining the dead
containers to verify that the container has already been run. This is fine
in most cases, but if the dead containers were garbage collected, kubelet may
falsely concluded that the pod has never been run. It would then try to restart
all the containers.

This change eliminates most of such possibilities by pre-filtering out the pods
in the final states before sending updates to per-pod workers.

### CODE REVIEW

**Comment by yujuhong** [comment]
This fixes #7250

Note that this PR handles most cases, but not all:
 - If the containers were removed immediately after they died, even generating correct status could fail.
 - If kubelet restarts before writing the terminated status for a pod to the apiserver, it could still restart a terminated pod (even though it'd not be reflected via the pod phases).

I think it's reasonable to assume that containers should only be garbage collected after a reasonable amount of time. As for the restart, we'd need a checkpoint of kubelet to handle that smoothly.

One upside of this PR is that kubelet would stop wasting resources on terminated pods (examining containers, generating statuses and sending them to the apiserver). It would also avoid (potential) post-mortem status modifications.

**Comment by yujuhong** [comment]
passed e2e

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1444,16 +1466,13 @@ func (kl *Kubelet) SyncPods(allPods []*api.Pod, podSyncTypes map[types.UID]metri
...
-		if ok && status.Phase == api.PodFailed {
-			continue
-		}
-		pods = append(pods, pod)
-	}
+
+	// Pod phase progresses monotonically. Once a pod has reach a final state,
+	// it should never leave irregardless of the restart policy. The statuses
+	// of such pods should not be changed, and there is no need to sync them.
+	pods := kl.FilterOutTerminatedPods(allPods)
```
Comment (dchen1107) [inline pkg/kubelet/kubelet.go]: Please also add TODO to explain those uncovered cases.

**Comment by dchen1107** [comment]
LGTM overall except one nit.

**Comment by vmarmol** [comment]
LGTM

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Done.

**Comment by dchen1107** [comment]
LGTM and will merge once it is green.

**Pull request merged**